lsp-mode.el

```diff
@@ -6065,17 +6069,25 @@ When UPDATE? is t force installation even if the server is present."
 
 ;; npm handling
 
+(defvar lsp--npm-dependency-path-cache nil
```

Comment:
is this cache really needed?

Reply:
Strictly speaking the cache isn't needed, when HOME is a local file system the redundant stat's for newly opened files aren't noticeable. In our environment we map HOME (~) for Linux, Windows, Mac to the same NFS / CIFS filer location. stat's against these can stall and do in general take much longer than a local file system. So, it does help in this scenario.

Customizing `lsp-server-install-dir` to a shared location didn't end up working for us. The reason is we want the "organization standard" languages servers in a shared read-only network location. People still need to be able to download non-organization standard servers for languages we haven't baked into our default emacs setup. The solution I took was to use `(puthash "server-name" "/net/location" lsp--npm-dependency-path-cache)` which lets us customize lsp without having to maintain our tweaks. This probably isn't the best way to state where "organization standard" language servers come from without having to modify `lsp-server-install-dir`. If you have something I could tryout, that would be nice.

Thanks

Reply:
Got it. But still, I do not think that the caching is part of npm provider. Imagine that we add several installers(yarn, pip, composert, etc). Each one of them will need caching for that particular use-case. The caching is a better fit for the `:system` provider.  But before jumping into implementing a cache for the system, I think that the following will do the job for you:

``` elisp
(with-eval-after-load 'lsp-clients
  (lsp-dependency typescript-language-server
    (:system "/home/kyoncho/.emacs.d/.cache/lsp/npm/typescript-language-server/bin/typescript-language-server")))
``` 

Note also the you could create providers optimized for your organization and register them in `lsp-deps-providers` and then do `(lsp-dependency ...`. (We could introduce a method `lsp-dependency-add-provider` to make it more consistent. 

Let me know what do you think.

Reply:
And one more thing that you may consider for your organization - lsp-mode supports running the language server in docker container over a local source tree. This may yield better performance and might be more convenient to manage the dependencies - e. g. everyone will have a consistent experience.

Reply:
I tried modifying `lsp--dependencies` via

            (ht-set lsp--dependencies 'typescript-language-server
                    (list `(:system ,tls-exe)))
            (ht-set lsp--dependencies 'typescript
                    (list `(:system ,ts-exe))))

and validated the `lsp--dependencies` paths for :system are correct. However, this doesn't work? This is stretching my knowledge of lisp. I'm not familiar with macros, ht, etc. thus having a little difficulty following the flow of the code.

Note, I can't call `lsp-dependency` because the paths to the executables are dynamically computed and are in a variable. Calling `lsp-dependency` doesn't expand the variable.

Nice to know lsp supports docker containers, but I don't think we need to use a docker container in our environment because we manage the locations of the servers, etc. and rely on central filers to update them on a regular clip. However, I need to investigate more to see if the containers are a better approach.

Reply:
> However, this doesn't work? 

:system uses executable-find which wont work for .js file I think. You may do `M-x trace-function RET executable-find RET` and track down the invocations of executable-find. 

> Note, I can't call lsp-dependency because the paths to the executables are dynamically computed and are in a variable. Calling lsp-dependency doesn't expand the variable.

We might replace the macro here with a function since it looks like it is causing more harm than good.

Reply:
Yes, `executable-find` when called from `lsp-package-path`, doesn't work on a *.js file, yet the tsserver requires a js file on Windows. This is because the typescript compiler ,tsserver, is invoked by the typescript-language-server and thus we pass the path to  tsserver.js to ensure typescript-language-server is able to spawn the tsserver. On UNIX, you can pass a path to a bash script that runs the tsserver and this works. However, on Windows if you pass a CMD script that runs the tsserver compiler, the typescript-language-server fails when spawning. The language server is called with the tsserver path and when passed a CMD file, when the typescript-language-server calls `cp.spawn(tsserverPath, args)`, it fails. If you pass the path to the tsserver.js things work on all platforms.

It seems to me that if I'm providing the location of the necessary items, that lsp shouldn't be doing `executable-find`, etc. It should just use what was specified.

Would it make sense to add a `defcustom lsp-package-locations` that lets one map package dependencies to absolute locations and this could be used to configure all language servers? This `defcustom` would be used by `lsp-package-path`. If this idea makes sense, I can try it out.

Another alternative is that I just put advice on `lsp-package-path`, but this is somewhat working around the missing capability to specify an absolute path to the packages, which to me seems like a useful feature for others.

Reply:
> It seems to me that if I'm providing the location of the necessary items, that lsp shouldn't be doing executable-find, etc. It should just use what was specified.

Haven't tested that but something like that should work:

```
(defvar lsp-deps-providers
  (list :npm (list :path #'lsp--npm-dependency-path
                   :install #'lsp--npm-dependency-download)
        :system (list :path #'lsp--system-path)))

(defun lsp--system-path (path)
  (or (executable-find path)
      (and (f-absolute? path)
           (f-exists? path)
           path)))
```
and the other change is to convert lsp-dependency from macro to function to allow params evaluation without doing tricks.

Reply:
You probably can use this:

```emacs-lisp
(defun lsp--system-path (path)
  (or (and (file-executable-p path)
           path)
      (executable-find path)))
```

Reply:
Thanks for the pointers. Sorry I've been a little slow in responding (been busy with other items). Hopefully, all is good now. I removed the cache code since it probably unique to our environment and I can manage this on our end.

Note, we can't use `executable-find` because some of the dependencies are not executables. For example the typescript compiler, tsserver.js is a JavaScript file. Passing the path to the tsserver.cmd file on Windows causes the typescript-language-server to fail. See comment above.

Reply:
> However, on Windows if you pass a CMD script that runs the tsserver compiler, the typescript-language-server fails when spawning. The language server is called with the tsserver path and when passed a CMD file, when the typescript-language-server calls `cp.spawn(tsserverPath, args)`, it fails. If you pass the path to the tsserver.js things work on all platforms.

Out of curious, which version of `typescript-language-server` are you using? I'm using the latest one and it seems fine for me to pass cmd file path as parameter of `--tsserver-path`.
Maybe that's the limitation of `cmd` in Windows that is not run well with network path.

Reply:
> I removed the cache code since it probably unique to our environment and I can manage this on our end.

AFAICS the latest version still has cache?

Reply:
Now it's really removed. Thanks for the catch.

Reply:
Regarding the tsserver-path issue. Here's the info on spawning child processes https://nodejs.org/api/child_process.html#child_process_spawning_bat_and_cmd_files_on_windows 

We have in typescript-language-server/lib/node_modules/typescript-language-server/lib/tsp-client.js, `cp.spawn(tsserverPath, args);` and it's not setup to invoke the cmd file. 

I'm using typescript-language-server version 0.4.0 with typescript compiler (tsserver) 3.7.4